### PR TITLE
Compositor & Scene: Simplify buffering

### PR DESCRIPTION
**Pull request by RAOF**
In various places we notify observers of how many buffers have changed when events happen.

Since we don't have any queues anymore, we only ever supply `1` as the number of buffers.

Remove some of the buffer counts and simplify `MultiThreadedCompositor` appropriately.

### CODE REVIEW

File: src/server/compositor/multi_threaded_compositor.cpp [hunk 1]
```diff
@@ -254,18 +210,15 @@ class CompositingFunctor
     std::shared_ptr<mc::DisplayBufferCompositorFactory> const compositor_factory;
     mg::DisplaySyncGroup& group;
     std::shared_ptr<mc::Scene> const scene;
+    std::binary_semaphore wakeup;
     bool running;
```
Comment (AlanGriffiths) [inline src/server/compositor/multi_threaded_compositor.cpp:214]: We're no longer guarding `running` with `run_mutex`. So we should make it `atomic<>`

**Review by AlanGriffiths — CHANGES_REQUESTED** [review]
I can't help feeling that we will want this back at some point. But I agree, "not yet".

However, I'm not yet convinced that the logic replacing the `running` bool/mutex/condition_variable is right

File: src/server/compositor/multi_threaded_compositor.cpp (hunk 1 above)
Comment (RAOF) [reply src/server/compositor/multi_threaded_compositor.cpp]: I'm *moderately* certain that the semaphore provides a sufficient barrier, but just using an `atomic<bool>` will definitely make it *obviously* correct.

**Comment by RAOF** [comment]
> I can't help feeling that we will want this back at some point. But I agree, "not yet".

Yeah, I've had a look at the proposed `wp-transaction`/`wp-commit-timing`/`wp-fifo` protocols. We will absolutely need *some* sort of queuing in the future, but I don't think what I'm removing here will be the right abstraction when we want to implement those anyway.

**Comment by RAOF** [comment]
@AlanGriffiths I've switched to `std::atomic<bool> running`. Any other concerns?

**Comment by AlanGriffiths** [comment]
> @AlanGriffiths I've switched to `std::atomic<bool> running`. Any other concerns?

Yeah, sorry I didn't get back to it yesterday - too many distractions. I need to get my head around the synchronization and convince myself it really works

File: src/server/compositor/multi_threaded_compositor.cpp (hunk 1 above)
Comment (AlanGriffiths) [reply src/server/compositor/multi_threaded_compositor.cpp]: I don't see that there was a sufficient barrier:

Thread A: `CompositingFunctor::operator()()` - acquires
Thread B: `CompositingFunctor::stop()` - running = false
Thread A: `CompositingFunctor::operator()()` - if (running) // undefined behavour: no guarantee running is one of true or false (even though it is on Intel, for example)

**Review by AlanGriffiths — CHANGES_REQUESTED** [review]

File: src/server/compositor/multi_threaded_compositor.cpp [hunk 2]
```diff
@@ -196,41 +164,30 @@ class CompositingFunctor
...
-        std::unique_lock lock{run_mutex};
-
-        if (num_frames > frames_scheduled)
-        {
-            frames_scheduled = num_frames;
-            lock.unlock();
-            run_cv.notify_one();
-        }
+        wakeup.try_acquire();
+        wakeup.release();
```
[the remark this answers is not in the document]
Comment (AlanGriffiths) [reply src/server/compositor/multi_threaded_compositor.cpp:170]: **Preconditions**

Both `update >= 0` and `update <= max() - counter` are true, where counter is the value of the internal counter. 

https://en.cppreference.com/w/cpp/thread/counting_semaphore/release

[the remark this answers is not in the document]
Comment (RAOF) [reply src/server/compositor/multi_threaded_compositor.cpp]: Oh, urgh, you're quite right. And it's worse; even *if* `try_acquire` doesn't spuriously fail, there's a race between the `try_acquire` and `release` that could result in two threads `release`ing at the same time.

I *know* there's an OS primitive for “sleep until this value changes”, damnit. Why doesn't C++ expose that‽

**Review by AlanGriffiths — APPROVED** [review]
OK. Cool! I had missed `atomic<>` growing `wait()`/`notify...()`. I like it!

**Pull request merged**